Feature/remember props

gswirski opened the pull request:
// this should be good to merge now

I couldn't test this on the train but figured some annotations might be worth sharing anyway.

libgphoto2 probably has this decently documented. I just have a nice puzzle while the internet is very spotty. :)

gswirski committed:
try to remember f number

gswirski committed:
more annotations

gswirski commented:
Potentially relevant: https://github.com/gphoto/libgphoto2/blob/bcd86bab811a0ca0ae907f8ed3b592e050a53c46/camlibs/ptp2/ptp.c#L3710

Potentially even more relevant: https://github.com/gphoto/libgphoto2/blob/560e15d3f4e51f1dcf161e9fbaac4922ecd51649/libgphoto2_port/vusb/vcamera.c#L1311

gswirski committed:
almost complete f number annotations

simonmitchell committed:
Documents focus modes. Makes sure app still works by adding in missing '+'

gswirski committed:
full support for f number updates

gswirski commented:
Data types for future reference:

```
	case 0x1:	return put_8bit_le (data, val->i8);
	case 0x2:	return put_8bit_le (data, val->u8);
	case 0x3:	return put_16bit_le (data, val->i16);
	case 0x4:	return put_16bit_le (data, val->u16);
	case 0x6:	return put_32bit_le (data, val->u32);
	case 0xffff:	return put_string (data, val->str);
```

simonmitchell committed:
Documents metering modes

simonmitchell committed:
Merge origin/feature/remember-props into feature/remember-props

gswirski commented:
@simonmitchell the app crashes after these commits.

(btw I wonder if some poor Sony oncall person was woken up by all of the crashes caused by me...)

gswirski commented on scripts/main-loop.js:
```diff
@@ -176,20 +176,27 @@ define(['./packet', './loop-factory', './data-factory', './device-prop-codes'],
                     "07 50" + // f number
                     "04 00" + // data type. uint16
-                    "01 01 ff ff" + // get/set (uint8), unknown (uint8), factory value (data type)
+                    // get (01 = available, 02 = unavailable), set (01 = available, 02 = unavailable)
```
this is brilliant!

gswirski commented on scripts/main-loop.js:
```diff
@@ -176,20 +176,27 @@ define(['./packet', './loop-factory', './data-factory', './device-prop-codes'],
                     "02" + // enumeration (1=range, 2=enum)
-                    "13 00" + // 19 values in array @simon: 1 array will be "available" and one will be "supported".
-                    "18 01 40 01 5e 01 90 01 c2 01 f4 01 30 02 76 02" + // list starts with 280 (F2.8)
+                    "13 00" + // 19 values in array @simon: 1 array will be "available" and one will be "supported" not able to confirm this.
```
You're probably right given it's how it used to work in the old HTTP API. Would be great to find an example where these two lists are not identical.

 I'm a bit surprised that shutter speed has 55 elements in both lists, even though we are in aperture-priority mode (`set` flag is set to `02` though).

simonmitchell replied:
Flash mode below has 5 in the second array and 3 in the first :) The ones missing from the first array are "Auto" and "Off"... I'm thinking that this means that the second array is supported, and the first is available. Will be interesting to see if there are any further examples throughout the data! It's a bit tricky to tell because if you disable values in either of them they disappear from Sony's UI. However if the first array is longer than the second Sony's parsing seems to break and the full configured UI isn't shown which makes me think my above assumption is correct!

simonmitchell committed:
Documents more props

simonmitchell commented:
> @simonmitchell the app crashes after these commits.
> 
> (btw I wonder if some poor Sony oncall person was woken up by all of the crashes caused by me...)

Hahahaaa quite possibly! They've only got themselves to blame though for not providing documentation and forcing us to reverse-engineer it!

simonmitchell committed:
Documents more props

simonmitchell commented:
@gswirski just a heads up for my plan for this.
1. Finish off dissecting the payload structure to pull out the “real world” representations of each prop type.
2. Once that is done, create a “camera” script/class/whatever which holds the server state and defines the allowed/supported values of each prop type as constants.
3. Create a method in packet.js which creates an individual segment of this data payload.
4. Create a method in packet.js (or elsewhere if you can think of a more appropriate place!) that creates the full payload.

simonmitchell committed:
Finishes documenting drive modes
Updates createEventPacket to send correct transaction ID
Server state for drive mode
Adds new data method to create readable big-endian hex from data

simonmitchell committed:
Adds scripts for representing a device prop and a camera object.
Starts moving returned payloads over to using objects and auto-generated payloads.

simonmitchell committed:
Moves still capture mode to camera.js

simonmitchell committed:
Moves a load more props to `camera.js`
